### PR TITLE
Auto-expand indices according to allocation filtering rules

### PR DESCRIPTION
**Pull request by ywelsch**
Honours allocation filtering rules when auto-expanding indices.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/allocation/FilteringAllocationIT.java [hunk 1]
```diff
@@ -97,6 +98,45 @@ public void testDecommissionNodeNoReplicas() {
...
     }
 
+    public void testAutoExpandReplicasToFilteredNodes() {
+        logger.info("--> starting 2 nodes");
+        List<String> nodesIds = internalCluster().startNodes(2);
+        final String node_0 = nodesIds.get(0);
+        final String node_1 = nodesIds.get(1);
+        assertThat(cluster().size(), equalTo(2));
+
+        logger.info("--> creating an index with no replicas");
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/allocation/FilteringAllocationIT.java]: nit: not no replicas

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
One comment, however, is that I don't think this API generalises to one that would address https://github.com/elastic/elasticsearch/issues/2869.

**Comment by ywelsch** [comment]
> One comment, however, is that I don't think this API generalises to one that would address #2869.

That was a non-goal here. I don't think that there's a general solution for #2869, so would like to focus on the use cases that we can solve.

Note that I've pushed https://github.com/elastic/elasticsearch/pull/48974/commits/972af8b75f2c3c73995124a547f479a4e24f344d to revise the comment in the docs about auto-expand replicas and shard allocation rules.

File: docs/reference/index-modules.asciidoc [hunk 2]
```diff
@@ -96,11 +96,12 @@ specific index module:
     Auto-expand the number of replicas based on the number of data nodes in the cluster.
     Set to a dash delimited lower and upper bound (e.g. `0-5`) or use `all`
     for the upper bound (e.g. `0-all`). Defaults to `false` (i.e. disabled).
-    Note that the auto-expanded number of replicas does not take any other allocation
-    rules into account, such as <<allocation-awareness,shard allocation awareness>>,
-    <<shard-allocation-filtering,filtering>> or <<allocation-total-shards,total shards per node>>,
-    and this can lead to the cluster health becoming `YELLOW` if the applicable rules
-    prevent all the replicas from being allocated.
+    Note that the auto-expanded number of replicas does only take
```
Comment (DaveCTurner) [inline docs/reference/index-modules.asciidoc]:
```suggestion
    Note that the auto-expanded number of replicas only takes
```

**Pull request merged**